File: ipylab/contents.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) ipylab contributors.
# Distributed under the terms of the Modified BSD License.

import typing as t
from uuid import uuid4
from ipywidgets import Widget, register
from traitlets import HasTraits, Unicode, Any, Instance, observe, Int, Bool
from ._frontend import module_name, module_version


@register
class ContentsManager(Widget):
    _model_name = Unicode("ContentsManagerModel").tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)

    _requests: t.Dict[str, t.Callable]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._requests = {}
        self.on_msg(self._on_frontend_msg)

    def _on_frontend_msg(self, _, content, buffers):
        _id = content.get("_id")
        callback = self._requests.pop(_id, None)
        if callback:
            callback(content)

    def get(self, path: str, content: t.Optional[bool] = None) -> "ContentsModel":
        _id = str(uuid4())
        self.send(
            {
                "_id": _id,
                "func": "get",
                "payload": {"path": path, "options": {"content": content}},
            }
        )

        model = ContentsModel(path=path, _contents_manager=self)

        self._requests[_id] = model._on_get

        return model

    def save(self, model: "ContentsModel"):
        _id = str(uuid4())
        self.send(
            {
                "_id": _id,
                "func": "save",
                "payload": {
                    "path": model.path,
                    "options": {
                        k: getattr(model, k)
                        for k in model.class_trait_names()
                        if not (
                            k.startswith("_") or k == "error" or k in Widget._traits
                        )
                    },
                },
            }
        )

        self._requests[_id] = model._on_get

        return model


@register
class ContentsModel(Widget):
    _model_name = Unicode("ContentsModelModel").tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)
    _contents_manager = Instance(ContentsManager)
    _syncing: t.Optional[bool]

    error = Unicode(allow_none=True)

    name = Unicode(allow_none=True)
    path = Unicode(allow_none=True).tag(sync=True)
    last_modified = Unicode(allow_none=True)
    created = Unicode(allow_none=True)
    format = Unicode(allow_none=True)
    mimetype = Unicode(allow_none=True)
    size = Int(allow_none=True)
    writeable = Bool(allow_none=True)
    type = Unicode(allow_none=True)
    content = Any(allow_none=True)

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.observe(self._on_content, "content")

    def _on_get(self, msg):
        if "error" in msg:
            self.error = msg["error"]
            return
        func = msg.get("func")
        self._syncing = True
        for key, value in msg.get("model", {}).items():
            if func == "save" and key == "content":
                continue
            setattr(self, key, value)
        self._syncing = False

    def _on_content(self, change) -> None:
        if self._syncing:
            return
        self._contents_manager.save(self)

File: src/widgets/contents.ts
// Copyright (c) ipylab contributors
// Distributed under the terms of the Modified BSD License.

import { WidgetModel } from '@jupyter-widgets/base';
import { Contents } from '@jupyterlab/services';

import { MODULE_NAME, MODULE_VERSION } from '../version';

export class ContentsManagerModel extends WidgetModel {
  /**
   * The default attributes.
   */
  defaults(): any {
    return {
      ...super.defaults(),
      _model_name: ContentsManagerModel.model_name,
      _model_module: ContentsManagerModel.model_module,
      _model_module_version: ContentsManagerModel.model_module_version
    };
  }

  /**
   * Initialize a ContentsManagerModel instance.
   *
   * @param attributes The base attributes.
   * @param options The initialization options.
   */
  initialize(attributes: any, options: any): void {
    super.initialize(attributes, options);
    this.on('msg:custom', this._onMessage.bind(this));
  }

  /**
   * Handle a custom message from the backend.
   *
   * @param msg The message to handle.
   */
  private async _onMessage(msg: any): Promise<void> {
    const { _id, payload, func } = msg;
    const { contentsManager } = ContentsManagerModel;

    let model: any;

    try {
      switch (func) {
        case 'get':
          model = await contentsManager.get(payload.path, payload.options);
          break;
        case 'save':
          model = await contentsManager.save(payload.path, payload.options);
          break;
        default:
          break;
      }
      this.send({ _id, func, model });
    } catch (err: any) {
      this.send({ _id, func, error: `${err}` });
    }
  }

  static model_name = 'ContentsManagerModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name: string = null;
  static view_module: string = null;
  static view_module_version = MODULE_VERSION;

  static contentsManager: Contents.IManager;
}

export class ContentsModelModel extends WidgetModel {
  /**
   * The default attributes.
   */
  defaults(): any {
    return {
      ...super.defaults(),
      contents: null,
      _model_name: ContentsModelModel.model_name,
      _model_module: ContentsModelModel.model_module,
      _model_module_version: ContentsModelModel.model_module_version
    };
  }

  static model_name = 'ContentsModelModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name: string = null;
  static view_module: string = null;
  static view_module_version = MODULE_VERSION;
}
